src/main/java/tutorlink/commandpackage/FindStudentCommand.java

```diff
@@ -1,42 +1,49 @@
+import tutorlink.exceptionspackage.IllegalValueException;
+import tutorlink.exceptionspackage.StudentNotFoundException;
+import tutorlink.exceptionspackage.TutorLinkException;
 import tutorlink.listpackage.StudentList;
 import tutorlink.resultpackage.CommandResult;
+import tutorlink.appstatepackage.AppState;
+import java.util.ArrayList;
 
 public class FindStudentCommand extends Command {
 
-    public static final String COMMAND_WORD = "find_student";
-    public static final String FORMAT_ERROR_MESSAGE = "Error, expected format: " + COMMAND_WORD
-            + " n/NAME {AND/OR} " + "i/MATRIC NUMBER";
-    public static final int PREFIX_INDEX = 2;
-    public static final String REGEX = "^find_student (i/(A\\d{7}[A-Z])?( n/[\\w\\d]+)?|n/[\\w\\d]" +
-            "+( i/(A\\d{7}[A-Z])?)?)( )?$";
+    public final String[] ARGUMENT_PREFIXES = {"i/", "n/"};
+    public final String COMMAND_WORD = "find_student";
 
-    protected static final String ERROR_MESSAGE = "Student (%s) not found in Student List";
+    private static final String ERROR_BOTH_NULL = "Error! Both parameters passed are null!";
+    private static final String DUPLICATE_MATRIC_NUMBER = "Error! Duplicate Matric Numbers found!";
 
-    private static final String SUCCESS_MESSAGE = "Found the following students:";
-    protected String name;
-    protected String matricNumber;
+//    public final String FORMAT_ERROR_MESSAGE = "Error, expected format: " + COMMAND_WORD
+//            + " n/NAME {AND/OR} " + "i/MATRIC NUMBER";
+//    public static final int PREFIX_INDEX = 2;
+//    public static final String REGEX = "^find_student (i/(A\\d{7}[A-Z])?( n/[\\w\\d]+)?|n/[\\w\\d]" +
+//            "+( i/(A\\d{7}[A-Z])?)?)( )?$";
 
-    public FindStudentCommand(String name, String matricNumber) {
-        this.name = name;
-        this.matricNumber = matricNumber;
-    }
+//    protected static final String ERROR_MESSAGE = "Student (%s) not found in Student List";
+//
+//    private static final String SUCCESS_MESSAGE = "Found the following students:";
 
     @Override
-    public CommandResult execute() {
-        StudentList filteredList = students.filterList(this.name, this.matricNumber);
-        if (filteredList.getNumberOfStudents() > 0) {
-            return new CommandResult(SUCCESS_MESSAGE, filteredList);
+    public CommandResult execute(AppState appstate, HashMap<String,String> hashmap) throws TutorLinkException {
+        String matricNumber = hashmap.get(ARGUMENT_PREFIXES[0]);
+        String name = hashmap.get(ARGUMENT_PREFIXES[1]);
+        StudentList students;
+        if(name == null && matricNumber == null) {
+            throw new IllegalValueException(ERROR_BOTH_NULL);
+        }
+        if(hashmap.containsKey(ARGUMENT_PREFIXES[0])) {
+            students = appstate.students.findStudentByMatricNumber(matricNumber);
+            if(students.getStudentArrayList().size() > 1) {
+                throw new DuplicateMatricNumberException(DUPLICATE_MATRIC_NUMBER);
```

Comment:
Checking for duplicate matric can be moved to adding student instead

Stopping the user from adding a duplicate student it's better than adding it than have no way to know what happen since we can't find the student afterward

Reply:
I agree that we should check if a duplicate student matric number exists when adding but shouldn't we also check during find as well to be safe?

Reply:
this might make it harder to troubleshoot if it does get pass the first check in addStudent.

When adding student, if the command somehow didn't know that they are duplicate but this one know, then there's no way for the user to find the duplicate id and delete it since, well this is find command. 

Obviously the user might know that this is the problem from the fact that the command throw, but it still a bad user experience

Reply:
Alright that works too, thanks!